(IA-2149) Make Runtime Labels Mutable in Patch endpoint

ehanna4 opened the pull request:
Allows for updating the labels on a runtime. Able to add or update a label or delete an existing one. Should not be able to change the default labels.

Added unit tests for the decoder and the for updateRuntimeRequest and updated Swagger

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ X] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

ehanna4 committed:
Initial add ability to update runtime labels

WIP - need to test

ehanna4 committed:
Initial add ability to update runtime labels

WIP - need to test

ehanna4 committed:
should work now

ehanna4 committed:
should work now

ehanna4 committed:
Should be working label upsert&delete, wip unit test

ehanna4 committed:
Should be working label upsert&delete, wip unit test

ehanna4 committed:
compiles but new tests fail

ehanna4 committed:
compiles but new tests fail

ehanna4 committed:
Very close - some cleanup still to do

- filter out empty string vals
- todo: remove prints && better way to check final map?

ehanna4 committed:
Very close - some cleanup still to do

- filter out empty string vals
- todo: remove prints && better way to check final map?

ehanna4 committed:
Cleaned up

ehanna4 committed:
Cleaned up

ehanna4 committed:
cleanup

ehanna4 committed:
Merge branch 'eh-add-Label-patch' of https://github.com/broadinstitute/leonardo into eh-add-Label-patch

ehanna4 committed:
Works to delete and update - added filtering for default values

todo: fix swagger and unit test for the decoder

ehanna4 committed:
Update RuntimeRoutes.scala

ehanna4 committed:
Working other than possibly the swagger schema section

ehanna4 committed:
Fully working according to manual tests

ehanna4 committed:
typos

ehanna4 committed:
Added test for decoder

ehanna4 committed:
Merge branch 'develop' into eh-add-Label-patch

Qi77Qi commented on automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/runtimes/RuntimePatchSpec.scala:
```diff
@@ -48,7 +48,9 @@ class RuntimePatchSpec extends FixtureAnyFreeSpec with LeonardoTestUtils with Le
-                                                    None)
+                                                    None,
+                                                    Map("" -> ""),
+                                                    List(""))
```
if you just need empty Map and empty List, you can do `Map.empty`, and `List.empty`

Qi77Qi commented on core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/runtimeModels.scala:
```diff
@@ -361,6 +361,16 @@ case class DefaultRuntimeLabels(runtimeName: RuntimeName,
     ).filterNot(_._2 == null)
 }
 
+object defaultLabelKeys {
```
if this is used for unit tests, I'd move it to https://github.com/DataBiosphere/leonardo/blob/develop/http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/CommonTestData.scala

rtitle replied:
I think this is intended to be a companion object, so should be `object DefaultRuntimeLabels {..}`

The intent is to return the default label _keys_. Could make that more explicit by:
```
object DefaultRuntimeLabels {
  // Creating a dummy instance to obtain the default label keys
  val defaultLabelKeys = DefaultRuntimeLabels(RuntimeName(""),
                                              GoogleProject(""),
                                              WorkbenchEmail(""),
                                              WorkbenchEmail(""),
                                              None,
                                              None,
                                              None).toMap.keySet
}
```

Unless anyone has a better way..

Qi77Qi reviewed (commented):
lgtm mostly...did you test it in a fiab?

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/http/runtimeRoutesModels.scala:
```diff
@@ -81,4 +81,6 @@ object UpdateRuntimeConfigRequest {
-                                      updateAutopauseThreshold: Option[FiniteDuration])
+                                      updateAutopauseThreshold: Option[FiniteDuration],
+                                      labelsToUpsert: LabelMap,
+                                      labelsToDelete: List[String])
```
Wonder if this should be a `Set[String]`

ehanna4 replied:
oh actually yeah that seems better👍

rtitle commented on http/src/main/resources/swagger/api-docs.yaml:
```diff
@@ -2013,6 +2013,8 @@ components:
               cloudService: gce
               machineType: n1-highmem-16
               diskSize: 500
+            labelsToUpsert: {new_label : value, old_label : new_value}
```
Is this valid yaml syntax? I usually run any swagger change through https://editor.swagger.io/

rtitle replied:
Looks like the swagger validator is happy 👍

ehanna4 replied:
yeah I've been using that validator check my changes

rtitle commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/api/RuntimeRoutes.scala:
```diff
@@ -378,7 +378,16 @@ object RuntimeRoutes {
+      ul <- x.downField("labelsToUpsert").as[Option[LabelMap]]
+      dl <- x.downField("labelsToDelete").as[Option[List[String]]]
+    } yield {
+      val labelsToDelete =
```
Could this be written as:
```
val labelsToDelete = dl.getOrElse(Set.empty).removedAll(defaultLabelKeys.defaultLabels.keySet)
val labelsToUpdate = ul.getOrElse(Map.empty).removedAll(defaultLabelKeys.defaultLabels.keySet).filter(_._2.nonEmpty)
```

rtitle replied:
Actually -- I'm thinking it might be better to return `BadRequest` with an error message to the caller if we receive any invalid labels. The caller is probably doing something wrong in this case -- better to fail explicitly than silently ignore the bad data.

ehanna4 replied:
Please see the latest commit - I have updated so that the decoder will now return a `Error: Bad Request` rather than filtering, as you suggested

rtitle commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala:
```diff
@@ -503,6 +503,19 @@ class RuntimeServiceInterp[F[_]: Parallel](config: RuntimeServiceConfig,
       else Async[F].unit
+
+      // Deleting labels
+      existingLabels <- labelQuery.getAllForResource(runtime.id, LabelResourceType.runtime).transaction
```
Looks like `existingLabels` is not used, can we remove this?

rtitle commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala:
```diff
@@ -503,6 +503,19 @@ class RuntimeServiceInterp[F[_]: Parallel](config: RuntimeServiceConfig,
+      // Deleting labels
+      existingLabels <- labelQuery.getAllForResource(runtime.id, LabelResourceType.runtime).transaction
+
+      _ <- req.labelsToDelete.traverse { label =>
```
It might be better to do all the label operations (upserts and deletes) in 1 transaction if possible.

For example we could define deletion by batch in `LabelComponent`:
```
  def deleteForResource(resourceId: Long, resourceType: LabelResourceType, keys: Set[String])(
    ): DBIO[Int] =
    labelQuery
      .filter(_.resourceId === resourceId)
      .filter(_.resourceType === resourceType)
      .filter(_.key inSetBind keys)
      .delete
```

(I think upserts still need to be done one by one.)

Then here we could do:
```
      dbOp = DBIO.seq(
        labelQuery.deleteForResource(runtime.id, LabelResourceType.runtime, req.labelsToDelete),
        req.labelsToUpsert.toList.traverse_ {
          case (k, v) =>
            labelQuery.save(runtime.id, LabelResourceType.runtime, k, v)
        }
      )
      _ <- dbOp.transaction
```

rtitle commented on http/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/api/HttpRoutesSpec.scala:
```diff
@@ -226,7 +226,9 @@ class HttpRoutesSpec
       Some(true),
-      Some(5.minutes)
+      Some(5.minutes),
+      Map("" -> ""),
```
Can do `Map.empty` and `List.empty`

ehanna4 committed:
PR Feedback

- use set instead of list
- throw error for bad label changes
- use appropriate empty() collections

Qi77Qi reviewed (approved):
lgtm 👍

ehanna4 committed:
Fix unit tests

ehanna4 committed:
Merge branch 'develop' into eh-add-Label-patch

rtitle commented on core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/JsonCodec.scala:
```diff
@@ -42,6 +42,9 @@ object JsonCodec {
     "num nodes must be >= 1",
     List.empty
   )
+  val deleteDefaultLabelsDecodingFailure = DecodingFailure("Default labels are not allowed to be deleted", List.empty)
```
👍 , thanks for adding this

rtitle commented on http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala:
```diff
@@ -503,6 +503,17 @@ class RuntimeServiceInterp[F[_]: Parallel](config: RuntimeServiceConfig,
         clusterQuery.updateAutopauseThreshold(runtime.id, updatedAutopauseThreshold, ctx.now).transaction.void
       else Async[F].unit
+
+      // Deleting labels
```
I think I had a comment about doing the label operations in the same DB transaction -- did that work?

ehanna4 replied:
Still working on that - I should have called my previous commit "PR feedback 1"

rtitle commented:
Decoding logic looks good, one comment in `RuntimeServiceInterp`. Also seems like there is a compilation error.

ehanna4 committed:
Final PR feedback and fix compilation issue

rtitle reviewed (approved):
:shipit: ! Nice work
